sui_programmability/framework/sources/Governance/ValidatorSet.move

```diff
@@ -0,0 +1,296 @@
+    /// If not found, returns (false, 0).
+    fun find_validator(validators: &vector<Validator>, validator_address: address): (bool, u64) {
+        let length = Vector::length(validators);
+        let i = 0;
+        while (i < length) {
+            let v = Vector::borrow(validators, i);
+            if (Validator::get_sui_address(v) == validator_address) {
+                return (true, i)
+            };
+            i = i + 1;
+        };
+        (false, 0)
+    }
+
+    /// Process the pending withdraw requests. For each pending request, the validator
+    /// is removed from `validators` and sent back to the address of the validator.
+    fun process_pending_withdraws(validators: &mut vector<Validator>, withdraw_list: &mut vector<u64>) {
+        sort_withdraw_list(withdraw_list);
+        while (!Vector::is_empty(withdraw_list)) {
+            let index = Vector::pop_back(withdraw_list);
+            let validator = Vector::remove(validators, index);
+            Validator::send_back(validator);
+        }
+    }
+
+    /// Process the pending new validators. They are simply inserted into `validators`.
+    fun process_pending_validators(validators: &mut vector<Validator>, pending_validators: &mut vector<Validator>) {
+        while (!Vector::is_empty(pending_validators)) {
+            let v = Vector::pop_back(pending_validators);
+            Vector::push_back(validators, v);
+        }
+    }
+
+    /// Given a list of all `validators`, each with stake updated properly, decide
+    /// who could become a validator in the next epoch.
+    /// A validator is chosen if its stake is in the range of min/max stake specified,
+    /// and is among the top `new_validator_set_size` in terms of stake amount.
+    fun decide_new_active_validators(
+        validators: &mut vector<Validator>,
+        new_validator_set_size: u64,
+        new_min_validator_stake: u64,
+        new_max_validator_stake: u64,
+    ) {
+        sort_validators_by_stake(validators);
+        let removed_validators = Vector::empty();
+        while (!Vector::is_empty(validators)) {
+            let v = Vector::borrow(validators, 0);
+            if (Validator::get_stake_amount(v) > new_max_validator_stake) {
+                Vector::push_back(&mut removed_validators, Vector::remove(validators, 0));
+            } else {
+                break
+            }
+        };
+        while (!Vector::is_empty(validators)) {
+            let length = Vector::length(validators);
+            let v = Vector::borrow(validators, length - 1);
+            if (Validator::get_stake_amount(v) < new_min_validator_stake) {
+                Vector::push_back(&mut removed_validators, Vector::pop_back(validators));
+            } else {
+                break
+            }
+        };
+        let length = Vector::length(validators);
+        // Note: This may cause epoch advancement to fail, if we don't
```

Comment:
I think ideally, the only way this should fail is if there are 0 validators. We should have a min, but I don't think a max is needed.